File: mcts.py
"""
AlphaZero MCTS for Italian draughts: a FLAT tree with cached positions, plus
leaf batching.

No `Node` object per node: the tree is a set of PARALLEL LISTS indexed by node
id (P, N, W, expanded, pos, term, children), and EVERY node CACHES its own
position. The descent therefore does NOT re-run `pos.play` at every simulation
(that was the dominant cost: copying the board): each `play` happens once per
edge, on the first visit. It also avoids allocating an object and a children
dict per node.

Leaf batching with virtual loss: one network evaluation per "wave" of
`batch_size` simulations.

run() returns an object whose `.children` maps each root Move to its visit
statistics (.N, .Q), the interface used by self-play and the tools.
"""
from __future__ import annotations
import math
import numpy as np


class _Stat:
    __slots__ = ("N", "Q")

    def __init__(self, N, Q):
        self.N = N
        self.Q = Q


class _Root:
    __slots__ = ("children",)

    def __init__(self, children):
        self.children = children


class MCTS:
    def __init__(self, evaluator, n_sims: int = 200, c_puct: float = 1.5,
                 dirichlet_alpha: float = 1.5, dirichlet_eps: float = 0.25,
                 batch_size: int = 8, virtual_loss: float = 1.0,
                 rng: np.random.Generator | None = None):
        self.evaluator = evaluator
        self.n_sims = n_sims
        self.c_puct = c_puct
        self.dirichlet_alpha = dirichlet_alpha
        self.dirichlet_eps = dirichlet_eps
        self.batch_size = max(1, batch_size)
        self.vloss = virtual_loss
        self.rng = rng or np.random.default_rng()

    def run(self, root_pos, add_noise: bool = True) -> _Root:
        # parallel lists indexed by node id (0 = root)
        self.P = [1.0]
        self.N = [0.0]
        self.W = [0.0]
        self.expanded = [False]
        self.pos = [root_pos]      # CACHED position (None until realized)
        self.term = [None]         # terminal value (float), or None if not terminal
        self.cids = [None]         # child ids
        self.cmoves = [None]       # children's Moves (parallel to cids)
        # key(position) -> (priors, value), for transpositions. The key is the
        # board plus the side to move and ignores the no-progress counter: a
        # transposition reached with a different counter reuses the first
        # evaluation (the C++ engine does the same).
        self._eval_cache = {}

        priors, _ = self.evaluator.evaluate_batch([root_pos])[0]
        self._expand(0, priors)
        if not self.cids[0]:
            return _Root({})
        if add_noise:
            self._add_dirichlet(0)

        sims = 0
        while sims < self.n_sims:
            b = min(self.batch_size, self.n_sims - sims)
            pending = []
            for _ in range(b):
                leaf, path = self._select_to_leaf()
                if self.term[leaf] is not None:
                    self._backup(path, self.term[leaf])
                else:
                    pending.append((leaf, path))
            sims += b
            if pending:
                # skip the forward pass for positions already evaluated (transpositions)
                miss = []
                for leaf, path in pending:
                    cached = self._eval_cache.get(self.pos[leaf].key())
                    if cached is not None:
                        priors, value = cached
                        if not self.expanded[leaf]:
                            self._expand(leaf, priors)
                        self._backup(path, value)
                    else:
                        miss.append((leaf, path))
                if miss:
                    res = self.evaluator.evaluate_batch([self.pos[l] for (l, _) in miss])
                    for (leaf, path), (priors, value) in zip(miss, res):
                        self._eval_cache[self.pos[leaf].key()] = (priors, value)
                        if not self.expanded[leaf]:
                            self._expand(leaf, priors)
                        self._backup(path, value)

        children = {mv: _Stat(self.N[cid],
                              self.W[cid] / self.N[cid] if self.N[cid] > 0 else 0.0)
                    for cid, mv in zip(self.cids[0], self.cmoves[0])}
        return _Root(children)

    # --- internals --------------------------------------------------------
    def _new_node(self, prior: float) -> int:
        self.P.append(prior); self.N.append(0.0); self.W.append(0.0)
        self.expanded.append(False); self.pos.append(None)
        self.term.append(None); self.cids.append(None); self.cmoves.append(None)
        return len(self.P) - 1

    def _expand(self, i: int, priors: dict) -> None:
        cids, cmoves = [], []
        for mv, p in priors.items():
            cids.append(self._new_node(p))
            cmoves.append(mv)
        self.cids[i] = cids
        self.cmoves[i] = cmoves
        self.expanded[i] = True

    def _add_dirichlet(self, i: int) -> None:
        cids = self.cids[i]
        noise = self.rng.dirichlet([self.dirichlet_alpha] * len(cids))
        eps = self.dirichlet_eps
        for cid, nz in zip(cids, noise):
            self.P[cid] = (1.0 - eps) * self.P[cid] + eps * float(nz)

    def _select_to_leaf(self):
        i = 0
        path = [0]
        c_puct = self.c_puct
        N, W, P = self.N, self.W, self.P
        while self.expanded[i]:
            cids = self.cids[i]
            ni = N[i]
            sqrt_n = math.sqrt(ni) if ni > 1.0 else 1.0
            best, bk = -1e30, 0
            for k in range(len(cids)):
                cid = cids[k]
                n = N[cid]
                q = -(W[cid] / n) if n > 0 else 0.0
                s = q + c_puct * P[cid] * sqrt_n / (1.0 + n)
                if s > best:
                    best, bk = s, k
            cid = cids[bk]
            if self.pos[cid] is None:                       # realize (one play per edge)
                child_pos = self.pos[i].play(self.cmoves[i][bk])
                self.pos[cid] = child_pos
                if child_pos.is_terminal():
                    self.term[cid] = child_pos.terminal_value(child_pos.turn)
            N[cid] += 1.0; W[cid] += self.vloss             # virtual loss
            path.append(cid)
            i = cid
        return i, path

    def _backup(self, path, value: float) -> None:
        vloss = self.vloss
        N, W = self.N, self.W
        for cid in path[1:]:                                # remove the virtual loss
            N[cid] -= 1.0; W[cid] -= vloss
        v = value
        for cid in reversed(path):
            N[cid] += 1.0; W[cid] += v; v = -v

File: web/engine/mcts.js
/**
 * AlphaZero MCTS for the browser: a FLAT tree with cached positions and leaf
 * batching. Port of mcts.py, including the two decisions that make it fast.
 *
 * NO NODE OBJECTS. The tree is a set of parallel arrays indexed by node id
 * (P, N, W, expanded, pos, term, children), and every node caches its own
 * position. The descent therefore never replays a move: each `play` happens
 * once per edge instead of once per simulation, and a `play` copies the board.
 *
 * LEAF BATCHING with a virtual loss: a wave of `batchSize` simulations collects
 * that many leaves, and the network sees them in ONE call. Do not expect much
 * from it here: measured in the browser, a batch of eight saved under a tenth
 * of the cost per position, because the forward pass dominates and the call
 * itself is cheap. It is kept because it costs nothing and because the search
 * is then the same shape as the Python one, which is what the parity test
 * compares.
 *
 * run() returns the root's children as a list of {move, N, Q}: the visit counts
 * are what the caller should play by, not the raw priors.
 */

/** Deterministic, seedable generator: the same game can be replayed exactly. */
export function mulberry32(seed) {
  let a = seed >>> 0;
  return function () {
    a |= 0; a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/** Gamma(alpha, 1) by Marsaglia-Tsang, the piece a Dirichlet draw is built from. */
function gamma(rand, alpha) {
  if (alpha < 1) {
    // Boost: Gamma(a) = Gamma(a+1) * U^(1/a), the standard way to keep alpha < 1
    // from degenerating. The default (1.5) never comes down this branch; a
    // caller asking for sharper noise would.
    return gamma(rand, alpha + 1) * Math.pow(rand() || Number.MIN_VALUE, 1 / alpha);
  }
  const d = alpha - 1 / 3;
  const c = 1 / Math.sqrt(9 * d);
  for (;;) {
    let x, v;
    do {
      // Box-Muller for a standard normal.
      const u1 = rand() || Number.MIN_VALUE, u2 = rand();
      x = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
      v = 1 + c * x;
    } while (v <= 0);
    v = v * v * v;
    const u = rand() || Number.MIN_VALUE;
    if (u < 1 - 0.0331 * x * x * x * x) return d * v;
    if (Math.log(u) < 0.5 * x * x + d * (1 - v + Math.log(v))) return d * v;
  }
}

function dirichlet(rand, alpha, n) {
  const g = [];
  let s = 0;
  for (let i = 0; i < n; i++) { const x = gamma(rand, alpha); g.push(x); s += x; }
  return g.map((x) => x / s);
}

export class MCTS {
  constructor(evaluator, {
    nSims = 200, cPuct = 1.5, dirichletAlpha = 1.5, dirichletEps = 0.25,
    batchSize = 8, virtualLoss = 1.0, rand = Math.random,
  } = {}) {
    this.evaluator = evaluator;
    this.nSims = nSims;
    this.cPuct = cPuct;
    this.dirichletAlpha = dirichletAlpha;
    this.dirichletEps = dirichletEps;
    this.batchSize = Math.max(1, batchSize);
    this.vloss = virtualLoss;
    this.rand = rand;
  }

  /**
   * Searches from `rootPos` and returns [{move, N, Q}], most visited first.
   * `addNoise` mixes Dirichlet noise into the root priors: it belongs to
   * self-play, not to a game against a person, so it defaults to off here.
   */
  async run(rootPos, { addNoise = false } = {}) {
    this.P = [1.0]; this.N = [0.0]; this.W = [0.0];
    this.expanded = [false];
    this.posOf = [rootPos];
    this.term = [null];
    this.cids = [null];
    this.cmoves = [null];
    // key(position) -> {priors, value}. The key is the board plus the side to
    // move and ignores the no-progress counter, exactly as the other two
    // implementations do, so a transposition reuses the first evaluation.
    this.evalCache = new Map();

    const [{ priors }] = await this.evaluator.evaluateBatch([rootPos]);
    this.expand(0, priors);
    if (!this.cids[0] || this.cids[0].length === 0) return [];
    if (addNoise) this.addDirichlet(0);

    let sims = 0;
    while (sims < this.nSims) {
      const b = Math.min(this.batchSize, this.nSims - sims);
      const pending = [];
      for (let k = 0; k < b; k++) {
        const { leaf, path } = this.selectToLeaf();
        if (this.term[leaf] !== null) this.backup(path, this.term[leaf]);
        else pending.push({ leaf, path });
      }
      sims += b;
      if (pending.length === 0) continue;

      const miss = [];
      for (const { leaf, path } of pending) {
        const cached = this.evalCache.get(this.posOf[leaf].key());
        if (cached) {
          if (!this.expanded[leaf]) this.expand(leaf, cached.priors);
          this.backup(path, cached.value);
        } else {
          miss.push({ leaf, path });
        }
      }
      if (miss.length > 0) {
        const res = await this.evaluator.evaluateBatch(miss.map((m) => this.posOf[m.leaf]));
        for (let i = 0; i < miss.length; i++) {
          const { leaf, path } = miss[i];
          const { priors, value } = res[i];
          this.evalCache.set(this.posOf[leaf].key(), { priors, value });
          if (!this.expanded[leaf]) this.expand(leaf, priors);
          this.backup(path, value);
        }
      }
    }

    const out = this.cids[0].map((cid, k) => ({
      move: this.cmoves[0][k],
      N: this.N[cid],
      Q: this.N[cid] > 0 ? this.W[cid] / this.N[cid] : 0.0,
    }));
    out.sort((a, b) => b.N - a.N);
    return out;
  }

  // --- internals -------------------------------------------------------
  newNode(prior) {
    this.P.push(prior); this.N.push(0.0); this.W.push(0.0);
    this.expanded.push(false); this.posOf.push(null);
    this.term.push(null); this.cids.push(null); this.cmoves.push(null);
    return this.P.length - 1;
  }

  expand(i, priors) {
    const cids = [], cmoves = [];
    for (const { move, prior } of priors) {
      cids.push(this.newNode(prior));
      cmoves.push(move);
    }
    this.cids[i] = cids;
    this.cmoves[i] = cmoves;
    this.expanded[i] = true;
  }

  addDirichlet(i) {
    const cids = this.cids[i];
    const noise = dirichlet(this.rand, this.dirichletAlpha, cids.length);
    const eps = this.dirichletEps;
    cids.forEach((cid, k) => { this.P[cid] = (1 - eps) * this.P[cid] + eps * noise[k]; });
  }

  selectToLeaf() {
    let i = 0;
    const path = [0];
    const { N, W, P, cPuct } = this;
    while (this.expanded[i]) {
      const cids = this.cids[i];
      const ni = N[i];
      const sqrtN = ni > 1.0 ? Math.sqrt(ni) : 1.0;
      let best = -1e30, bk = 0;
      for (let k = 0; k < cids.length; k++) {
        const cid = cids[k];
        const n = N[cid];
        const q = n > 0 ? -(W[cid] / n) : 0.0;
        const s = q + cPuct * P[cid] * sqrtN / (1.0 + n);
        if (s > best) { best = s; bk = k; }
      }
      const cid = cids[bk];
      if (this.posOf[cid] === null) {            // realize the edge, once
        const childPos = this.posOf[i].play(this.cmoves[i][bk]);
        this.posOf[cid] = childPos;
        if (childPos.isTerminal()) this.term[cid] = childPos.terminalValue(childPos.turn);
      }
      N[cid] += 1.0; W[cid] += this.vloss;       // virtual loss
      path.push(cid);
      i = cid;
    }
    return { leaf: i, path };
  }

  backup(path, value) {
    const { N, W, vloss } = this;
    for (let k = 1; k < path.length; k++) { N[path[k]] -= 1.0; W[path[k]] -= vloss; }
    let v = value;
    for (let k = path.length - 1; k >= 0; k--) {
      N[path[k]] += 1.0; W[path[k]] += v; v = -v;
    }
  }
}
